dali/kernels/reduce/mean_stddev_gpu_impl.cuh

```diff
@@ -0,0 +1,386 @@
+  using PreprocessorBank = VariancePreprocessor<non_reduced_dims, Mean>;
+
+  void InitMean(const InListGPU<Mean> &mean) {
+    mean_ = reshape(mean, This().SimplifiedOutputShape(), true);
+  }
+
+  Preprocessor GetPreprocessorImpl() const {
+    assert(This().SimplifiedOutputShape().num_elements() == 1);
+    return Preprocessor { mean_.data[0] };
+  }
+
+  Preprocessor *GetPreprocessorsImpl(WorkArea &wa) const {
+    int n = This().SimplifiedInputShape().num_samples();
+    assert(This().SimplifiedOutputShape().num_elements() ==
+           This().SimplifiedOutputShape().num_samples());
+    Preprocessor *pp = wa.ParamBuffer<Preprocessor>(n);
+    for (int i = 0; i < n; i++) {
+      int o = This().ReduceBatch() ? 0 : i;
+      pp[i] = { mean_.data[o] };
+    }
+    return pp;
+  }
+
+  PreprocessorBank<1> *
+  GetPreprocessorBanks(WorkArea &wa, int axis, int_const<1>) const {
+    using Bank = PreprocessorBank<1>;
+    int n = This().SimplifiedInputShape().num_samples();
+    Bank *banks = wa.ParamBuffer<Bank>(n);
+
+    for (int i = 0; i < n; i++) {
+      int o = This().ReduceBatch() ? 0 : i;
+      auto shape = This().SimplifiedOutputShape().tensor_shape_span(o);
+      auto &bank = banks[i];
+      bank.mean = mean_.data[o];
+      bank.stride[0] = volume(shape.begin() + axis, shape.end());  // outer stride
+    }
+    return banks;
+  }
+
+  PreprocessorBank<2> *
+  GetPreprocessorBanks(WorkArea &wa, int axis, int_const<2>) const {
+    using Bank = PreprocessorBank<2>;
+    int n = This().SimplifiedInputShape().num_samples();
+    Bank *banks = wa.ParamBuffer<Bank>(n);
+
+    SmallVector<int, 6> remaining_axes;
+    for (int a : This().SimplifiedAxes())
+      if (a > axis)
+        remaining_axes.push_back(a - axis - 1);
+    int mask = to_bit_mask(remaining_axes);
+
+    for (int i = 0; i < n; i++) {
+      int o = This().ReduceBatch() ? 0 : i;
+      auto &bank = banks[i];
+      auto in_shape = This().SimplifiedInputShape().tensor_shape_span(i);
+      auto out_shape = This().SimplifiedOutputShape().tensor_shape_span(o);
+      auto inner_shape = span<const int64_t>(in_shape.begin() + axis + 1, in_shape.end());
+      bank.mean = mean_.data[o];
+      bank.stride[0] = volume(out_shape.begin() + axis, out_shape.end());  // outer stride
+      bank.stride[1] = 1;  // inner stride, always 1?
+      bank.inner_dims = DropDims(inner_shape, mask);  // reindexing, if necessary
+    }
+    return banks;
+  }
+
+  template <int non_reduced_dims>
+  PreprocessorBank<non_reduced_dims> *
+  GetPreprocessorBanksImpl(WorkArea &wa, int axis, int_const<non_reduced_dims> nrd) const {
+    return GetPreprocessorBanks(wa, axis, nrd);
+  }
+};
+
+template <typename Out, typename Scale>
+struct ScaleSqrtConvert {
+  Scale scale = 1;
+
+  template <typename T>
+  DALI_HOST_DEV Out operator()(T x) const {
+    return ConvertSat<Out>(sqrt(x * scale));
+  }
+};
+
+template <typename Out, typename In, typename Actual>
+class RootMeanImplBase {
```

Comment:
This is the same as MeanImplBase, but uses the ScaleSqrtConvert instead of ScaleAndConvert, right?

I'm thinking if taking some common parts into ElementwiseScaleAndSomething would have any benefit for tersness of this file, or is it better to keep separate.
Maybe keep it simpler.

Reply:
It could be sth like `MeanBaseImpl<typename Out, typename In, typename Actual, typename Potprocessor>`.